pyfr/plugins/ascent.py

```diff
@@ -0,0 +1,387 @@
+        for k in self.cfg.items(cfgsect, prefix='field-'):
+            field = k.removeprefix('field-')
+
+            if field in self._fields_write:
+                raise KeyError(f"Field '{field}' already exists")
+            self._fields_write.add(field)
+
+            exprs = self.cfg.getexpr(cfgsect, k, subs=cons)
+            self._exprs.append((field, f'fields/{field}/values', exprs))
+
+    def _init_gradients(self):
+        # Determine what gradients, if any, are required
+        g_pnames = set()
+        for field, path, expr in self._exprs:
+            g_pnames.update(re.findall(r'\bgrad_(.+?)_[xyz]\b', expr))
+
+        privarmap = self.elementscls.privarmap[self.ndims]
+        self._gradpinfo = [(pname, privarmap.index(pname))
+                           for pname in g_pnames]
+
+    def _init_pipelines(self, cfgsect):
+        pipel_cfg = {}
+        self.pipelines = pl = ConduitNode(self.conduit)
+
+        for k in self.cfg.items(cfgsect, prefix='pipeline-'):
+            pn = k.removeprefix('pipeline-')
+            cfg = self.cfg.getliteral(cfgsect, k)
+            pipel_cfg[pn] = cfg = [cfg] if isinstance(cfg, dict) else cfg
+
+            for j, filter in enumerate(cfg):
+                params = ConduitNode(self.conduit)
+
+                pl[f'pl_{pn}/f{j}/type'] = ptype = filter.pop('type')
+                for kf, vf in filter.items():
+                    if kf == 'output-name':
+                        if vf in self._fields_write:
+                            raise KeyError(f"Output name '{vf}' already used")
+                        self._fields_write.add(vf)
+                    elif kf == 'field':
+                        self._fields_read.add(vf)
+
+                    params[kf.replace('-', '_')] = vf
+
+                if ptype in self.v_filter:
+                    self._fields_read.add('velocity')
+
+                pl[f'pl_{pn}/f{j}/params'] = params
+
+    def _init_scenes(self, cfgsect):
+        self._scene_cfg = scene_cfg = {}
+        self.scenes = ConduitNode(self.conduit)
+
+        for k in self.cfg.items(cfgsect, prefix='scene-'):
+            sn = k.removeprefix('scene-')
+            scene_cfg[sn] = cfg = self.cfg.getliteral(cfgsect, k)
+
+            if cfg['type'] != 'mesh':
+                field = cfg['field']
+                self._fields_read.add(field)
+
+            for kc, vc in cfg.items():
+                if kc == 'pipeline':
+                    self.scenes[f's_{sn}/plots/p1/pipeline'] = f'pl_{vc}'
+                elif kc.startswith('render-'):
+                    rname = kc.removeprefix('render-')
+                    path = f's_{sn}/renders/r{rname}'
+
+                    self._render_options(path, vc)
+                elif kc != 'image-name':
+                    key = kc.replace('-', '_')
+                    self.scenes[f's_{sn}/plots/p1/{key}'] = vc
+
+            # If no render options then set image name
+            if not any(kc.startswith('render-') for kc in cfg):
```

Comment:
Is it worth forcing a set of render options? Cleans up the code and makes documentation a little easier since things are more explicit.